### PR TITLE
Convert engine events to use EventEmitter2; wait for promises to resolve

### PR DESCRIPTION
**Pull request by CacheControl**
Uses the `emitAsync()` feature of EventEmitter2 to ensure promises returned to emitted events resolve before engine continues

### CODE REVIEW

File: src/engine.js [hunk 1]
```diff
@@ -191,11 +192,13 @@ class Engine extends EventEmitter {
       return rule.evaluate(almanac).then((ruleResult) => {
         debug(`engine::run ruleResult:${ruleResult.result}`)
         if (ruleResult.result) {
-          this.emit('success', rule.event, almanac, ruleResult)
-          this.emit(rule.event.type, rule.event.params, almanac, ruleResult)
-          almanac.factValue('success-events', { event: rule.event })
+          almanac.factValue('success-events', { event: ruleResult.event })
+          return Promise.all([
+            this.emitAsync('success', ruleResult.event, almanac, ruleResult),
+            this.emitAsync(ruleResult.event.type, ruleResult.event.params, almanac, ruleResult)
```
Comment (naveen-shetty) [inline src/engine.js:198]: In the current implementation `success` & `ruleResult.event.type` is triggered sequentially. This change is triggering them concurrently. Not a problem for us but just wanted to check if it is intentional.

File: src/engine.js [hunk 2]
```diff
@@ -191,11 +192,13 @@ class Engine extends EventEmitter {
       return rule.evaluate(almanac).then((ruleResult) => {
         debug(`engine::run ruleResult:${ruleResult.result}`)
         if (ruleResult.result) {
-          this.emit('success', rule.event, almanac, ruleResult)
-          this.emit(rule.event.type, rule.event.params, almanac, ruleResult)
-          almanac.factValue('success-events', { event: rule.event })
+          almanac.factValue('success-events', { event: ruleResult.event })
```
Comment (naveen-shetty) [inline src/engine.js:195]: As `almanac.factValue` returns a promise, Shouldn't this go inside `Promise.all` array as well?

**Review by naveen-shetty — COMMENTED** [review]
Thanks @CacheControl for making this change.  Just a couple of things I noticed.

File: src/engine.js (hunk 2 above)
Comment (CacheControl) [reply src/engine.js:195]: yikes, that promise handling has been broken for a while. good catch!

File: src/engine.js (hunk 1 above)
Comment (CacheControl) [reply src/engine.js:198]: another good call out! I can't see a good reason reason to deviate from the previous behavior; I've updated this to fire these in series, like it did before.

**Comment by CacheControl** [comment]
Thanks for the code review @naveen-shetty ! I've tweaked a few things and added docs/examples, but from a consumer's perspective the functionality hasn't changed since the 6.0.0-alpha-3 tag - feel free to use that until I release 6.0.0. I'm merging this into the `next-major` branch for the time being.

**Comment by naveen-shetty** [comment]
@CacheControl Thank you very much for this change & also addressing my comments. I appreciate it. 

Any idea when you would be releasing  `6.0.0` version?

**Pull request merged**